---
title: "2) Elucidating Spatial Cell Composition of Neuroblastoma (Group Project)"
start_date: "2024-01-01"
end_date: "2024-03-01"
excerpt: "<br/><img src='/images/cisbio_1.png'> \n \n <strong>Figure 2.1:</strong> Initial, Single-cell RNA-Seq Workflow of the project. \n <br/><img src='/images/cisbio_2.png'> \n \n <strong>Figure 2.2:</strong> Secondary, Single-cell Spatial RNA-seq Workflow of the project. Please click the title for more detailed information on results and outcomes."

collection: portfolio
aims:     
- Meta-analyse various single-cell studies of neuroblastoma, building a single-cell atlas that spans tumour cell diversity (Figure 2.1)
- Identify cell populations and tissue heterogeneity in single cell spatial transcriptomics (SCST) data using transfer learning from the single-cell reference. 
- Validate, test and scrutinise available tools for exploratory analysis of novel SCST data, including fine tuning and adpatation of a novel, multi-modal deep learning clustering approach (SiGra) (Figure 2.2)

skills:
  - task: "Single-Cell RNA-Seq Data Integration + Analysis"
    tools:
      - "Python: scanpy, rpy2, anndata2ri, integration(scanorama, scvi-tools), monocle, palantir " 
      - "R: seurat, integration (harmony, rPCA CCA, BBKNN), CONICSmat, CellChat"
  - task: "Single-Cell Spatial RNA-Seq Data Analysis"
    tools:
      - "squidpy, transfer learning (cell2location, singleR, Seurat, RCTD, scarches), nichedb, "
  - task: "Deep Learning (SiGRA Modification)"
    tools:
      - "torchvision, matplotlib(v2.1.1), torch, seaborn, tqdm, scikit_learn, torch_geometric, keras, optuna, weights and biases (wandb), xgboost"
  - task: "Shell Computing"
    tools:
      - Unix:, Git (init, clone, add, commit, status), pull, push, branch, merge), ssh, High Performance Computing (HPC), nohup, rsync, Slurm (sbatch), module, chmod (permissions) 
---

Responsibilities:
- 4-month Undergraduate Research Project (KCL Biochemistry BSc): ‘Integrative Multi-omics Analysis of Human Carotid Plaque Data Reveals Protective Smooth Muscle Cell Phenotypes’. - This project meta-analysed 3 scRNA-seq datasets to build one of the first comprehensive single-cell transcriptomic atlases of human carotid artery atherosclerosis, extract smooth muscle cell populations and interpret their different phenotypes with transcriptomic and proteomics data through clinically translatable multi-omics analysis.
- 3-month return offer for part-time work given the quality of research. 
- The goal of this internship was to contribute to the lab’s publication and to write my own original research manuscript. To create my own manuscript, results had to be validated through further analysis.

Key Goals:
- R/Python scripting to benchmark and integrate single-cell sequencing data: 70,000 carefully selected cells from datasets uploaded to various public repositories
- K-Means and Hierarchical cell clustering procedures 
- Differential gene expression analysis and analysis of proteomic signatures for calcification, sex differences, symptoms and location.
- Transcriptomics/proteomics data interpretation of 200+ original patient plaque samples.
- Validation of results: Spatial RNA-seq, Cell-Cell Communication analysis, Bulk Deconvolution techniques. 

Achievement:
- Co-Author for an original research manuscript (7% acceptance rate) published in Circulation (AHA Journals, Vol 133): ‘Proteomic Atlas of Atherosclerosis: The Contribution of Proteoglycans to Sex Differences, Plaque Phenotypes, and Outcomes’
¬- Collaboration between the Medical University of Vienna and KCL

<details>
  <summary>Click for more info</summary>
  
  - **Project Details**: This section provides detailed insights into the implementation process, challenges faced, and results obtained.
  - **Paper Reference**: "Reward Constrained Policy Optimization" by Tessler et al. can be accessed [here](https://openreview.net/pdf?id=SkfrvsA9FX).
  - **GitHub Code**: The complete code for this project is available on [GitHub](https://github.com/sudo-Boris/stable-baselines3).
  - **Article Submission**: Learn more about the theory and results of RCPO in the submitted article [here](https://iclr-blogposts.github.io/staging/blog/2023/Adaptive-Reward-Penalty-in-Safe-Reinforcement-Learning/).

</details>
