---
title: "Multi-Omics Pathway-based Data Integration Using Machine Learning and Large Language Models (LLMs)"
start_date: "2024-03-01"
end_date: "2024-10-01"
excerpt: "<br/><img src='/images/multi_omics.png'> \n \n <strong>Figure 1:</strong> General Workflow of the project. Please click the title for more detailed information on results and outcomes."
collection: portfolio
aims:     
-  Use LLM-based semantic searching to map more metabolite names to IDs, widening the scope of pathway analysis (Figure 1A)
- Apply novel pathway databases augmented with metabolites using deep learning, facilitating more comprehensive pathway mapping and attaining more robust biological predictions (Figure 1B)
- Develop an extension to PathIntegrate which will provide an unsupervised utility for pathway-based multivariate analysis that can be benchmarked with synthetic data simulations (Figure 1C)
skills:
  - task: "Data Processing, Analysis and Databases"
    tools:
      - "pandas, duckdb SQL, fancyimpute, networkx, missforest, statsmodels, plotly, gseapy, matplotlib, seaborn, missforest, scipy"
  - task: "Machine Learning + Web App"
    tools:
      - "leidenalg, streamlit (HTML + CSS), elasticsearch, sentence_transformers, HuggingFace, base64, mbpls"
      - "sklearn: metrics (f1_score, precision_score, recall_score, roc_auc_score, roc_curve, confusion_matrix), model_selection (train_test_split, cross_val_score, GridSearchCV), pipeline (Pipeline), preprocessing (StandardScalar), linear_model (LogisticRegression), decomposition (PCA), manifold (TSNE)"
  - task: "Miscellaneous"
    tools:
      - "response, requests, OpenAI, urllib, igraph, json, tdqm, warnings, SLURM"
  - task: "Bioinformatics Tools"
    tools:
      - "KEGG API, Reactome API, CytoScape, ssPA, GSEA, MOFA2, pathintegrate, iPATH3"
---

Responsibilities:
- 4-month Undergraduate Research Project (KCL Biochemistry BSc): ‘Integrative Multi-omics Analysis of Human Carotid Plaque Data Reveals Protective Smooth Muscle Cell Phenotypes’. - This project meta-analysed 3 scRNA-seq datasets to build one of the first comprehensive single-cell transcriptomic atlases of human carotid artery atherosclerosis, extract smooth muscle cell populations and interpret their different phenotypes with transcriptomic and proteomics data through clinically translatable multi-omics analysis.
- 3-month return offer for part-time work given the quality of research. 
- The goal of this internship was to contribute to the lab’s publication and to write my own original research manuscript. To create my own manuscript, results had to be validated through further analysis.

Key Goals:
- R/Python scripting to benchmark and integrate single-cell sequencing data: 70,000 carefully selected cells from datasets uploaded to various public repositories
- K-Means and Hierarchical cell clustering procedures 
- Differential gene expression analysis and analysis of proteomic signatures for calcification, sex differences, symptoms and location.
- Transcriptomics/proteomics data interpretation of 200+ original patient plaque samples.
- Validation of results: Spatial RNA-seq, Cell-Cell Communication analysis, Bulk Deconvolution techniques. 

Achievement:
- Co-Author for an original research manuscript (7% acceptance rate) published in Circulation (AHA Journals, Vol 133): ‘Proteomic Atlas of Atherosclerosis: The Contribution of Proteoglycans to Sex Differences, Plaque Phenotypes, and Outcomes’
¬- Collaboration between the Medical University of Vienna and KCL

<details>
  <summary>Click for more info</summary>
  
  - **Project Details**: This section provides detailed insights into the implementation process, challenges faced, and results obtained.
  - **Paper Reference**: "Reward Constrained Policy Optimization" by Tessler et al. can be accessed [here](https://openreview.net/pdf?id=SkfrvsA9FX).
  - **GitHub Code**: The complete code for this project is available on [GitHub](https://github.com/sudo-Boris/stable-baselines3).
  - **Article Submission**: Learn more about the theory and results of RCPO in the submitted article [here](https://iclr-blogposts.github.io/staging/blog/2023/Adaptive-Reward-Penalty-in-Safe-Reinforcement-Learning/).

</details>
